guardian/support-service-lambdas
# ops: send notification email to SRE team if a message appears in the DLQ

## What does this change?

- Create SNS topic `single-contribution-salesforce-writes-topic`
- Create subscription to SRE email supporter.revenue.engine@guardian.co.uk
- Create CloudWatch alarm `single-contribution-salesforce-writes-alarm`
  - If there is at least a message in the DLQ
  - Email to be sent every 24 hours if message is still in DLQ
  - Actions enabled only on PROD
- Update C4 model diagram and images

[**Trello Card**](https://trello.com/c/oUbLuPf7/367-single-contribution-record-add-alarms-if-messages-appear-in-dead-letter-queue)

## Why are you doing this?

This is part of the epic [Create a single contribution record in Salesforce via support-payment-api](https://trello.com/c/qibQVaR2/295-create-a-single-contribution-record-in-salesforce-via-support-frontend-payment-api).

Previous work:

1. [feat: create single-contribution-salesforce-writes project with SQS queues to handle single contribution messages](https://github.com/guardian/support-service-lambdas/pull/2059)
2. [ops: add rule to acquisition bus to send payment-api messages to the single-contribution-salesforce-writes-queue](https://github.com/guardian/support-service-lambdas/pull/2066)
3. [feat: add fields (postalCode, state, email) to AcquisitionDataRow](https://github.com/guardian/support-frontend/pull/5345)
4. [feat: write lambda polling SQS and creating a single contribution record in Salesforce](https://github.com/guardian/support-service-lambdas/pull/2080)
5. [fix: make state and postal code optional fields](https://github.com/guardian/support-service-lambdas/pull/2089)
6. [fix: make identityId optional](https://github.com/guardian/support-service-lambdas/pull/2101)
7. [docs: add README and C4 diagrams](https://github.com/guardian/support-service-lambdas/pull/2102)

## Architecture (C4 Model)

![SystemLandscape](https://github.com/guardian/support-service-lambdas/assets/39066365/5ee555c1-cd4c-40fa-b0e2-1cb1b40dca50)

![SystemContainer](https://github.com/guardian/support-service-lambdas/assets/39066365/dc02923d-faa0-4fa0-8945-cef9d9ff5c76)

## How to test

1. Send a default message to the [CODE queue](https://eu-west-1.console.aws.amazon.com/sqs/v3/home?region=eu-west-1#/queues/https%3A%2F%2Fsqs.eu-west-1.amazonaws.com%2F865473395570%2Fsingle-contribution-salesforce-writes-queue-CODE)
2. Check the [alarm](https://eu-west-1.console.aws.amazon.com/cloudwatch/home?region=eu-west-1#alarmsV2:alarm/CODE$3A+Failed+to+sync+single+contribution+to+Salesforce?) status is "In alarm"

## How can we measure success?

If a single contribution event is not processed correctly, we should receive a notification email.
